k-distribution/k-tutorial/1_basic/03_parsing/README.md

```diff
@@ -1 +1,627 @@
+into a type of production called a **bracket**. A bracket production in K
+is any production with the `bracket` attribute. It is required that such a
+production only have a single non-terminal, and the sort of the production
+must equal the sort of that non-terminal. However, K does not otherwise
+impose restrictions on the grammar the user provides for a bracket. With that
+being said, the most common type of bracket is one in which a non-terminal
+is surrounded by terminals representing some type of bracket such as
+`()`, `[]`, `{}`, `<>`, etc. For example, we can define the most common
+type of bracket, the type used by the vast majority of programming languages,
+quite simply.
+
+Consider the following modified definition, which we will save to
+`lesson-03-d.k`:
+
+```k
+module LESSON-03-D
+
+  syntax Boolean ::= "true" | "false"
+                   | "(" Boolean ")" [bracket]
+                   | "!" Boolean [function]
+                   | Boolean "&&" Boolean [function]
+                   | Boolean "^" Boolean [function]
+                   | Boolean "||" Boolean [function]
+
+endmodule
+```
+
+In this definition, if the user does not explicitly define parentheses, the
+grammar remains ambiguous and K's just-in-time parser will report an error.
+However, you are now able to parse more complex programs by means of explicitly
+grouping subterms with the bracket we have just defined.
+
+Consider `and-or-left.bool`:
+
+```
+(true && false) || false
+```
+
+Now consider `and-or-right.bool`:
+
+```
+true && (false || false)
+```
+
+If you parse these programs with `kast`, you will once again get a single
+unique AST with no error. If you look, you may spot that the bracket itself
+does not appear in the AST. In fact, this is a property unique to brackets:
+productions with the bracket attribute are not represented in the parsed AST
+of a term, and the child of the bracket is folded immediately into the parent
+term. This is the reason for the requirement that a bracket production have
+a single non-terminal of the same sort as the production itself.
+
+## Priority blocks
+
+Of course, in practice, very few formal languages outside the domain of
+natural language processing are ambiguous. The main reason for this is that
+parsing unambiguous languages is asymptotically faster than parsing ambiguous
+languages. Programming language designers instead prefer to use a notion of
+operator precedence and associativity to make expression grammars unambiguous
+by automatically choosing certain parses over other parses in cases where the
+grammar would otherwise be ambiguous.
+
+While it is sometimes possible to explicitly rewrite the grammar to remove
+these parses, because K's grammar specification and AST generation are
+inextricably linked, this is generally discouraged. Instead, we use the
+approach of explicitly expressing the relative precedence of different
+operators in different situations in order to resolve the ambiguity.
+
```

Comment:
This block has the tone of "contradicting" the previous block of text, even though it agrees with it. Doesn't seem to add much either, so I would just remove it.

```suggestion
```

Reply:
This paragraph actually does serve an important role. It explains to anyone who is previously familiar with other parser generators an important way in which parser generation in K is usually done differently than they might be used to.